lld/MachO/ObjC.cpp

```diff
@@ -186,13 +186,23 @@ ObjcCategoryChecker::ObjcCategoryChecker()
       roClassLayout(target->wordSize), listHeaderLayout(target->wordSize),
       methodLayout(target->wordSize) {}
 
-// \p r must point to an offset within a cstring section.
+// \p r must point to an offset within a CStringInputSection or a
+// ConcatInputSection
 static StringRef getReferentString(const Reloc &r) {
   if (auto *isec = r.referent.dyn_cast<InputSection *>())
     return cast<CStringInputSection>(isec)->getStringRefAtOffset(r.addend);
   auto *sym = cast<Defined>(r.referent.get<Symbol *>());
-  return cast<CStringInputSection>(sym->isec())
-      ->getStringRefAtOffset(sym->value + r.addend);
+  if (auto *s = dyn_cast_or_null<CStringInputSection>(sym->isec()))
+    return s->getStringRefAtOffset(sym->value + r.addend);
+
+  if (isa<ConcatInputSection>(sym->isec())) {
+    auto paddedData = sym->isec()->data.slice(sym->value);
+    uint32_t len = strnlen((const char *)paddedData.data(), paddedData.size());
+    auto strData = paddedData.slice(0, len);
+    return StringRef((const char *)strData.data(), strData.size());
```

Comment:
We should be able to use `len` directly. Also, do we need the `(const char *)` casts? I thought that was already the type of `data()`.

```suggestion
    return StringRef((const char *)strData.data(), len);
```

Reply:
The casts are necessary because:
```
no known conversion from 'const unsigned char *' to 'const char *'
```